### PR TITLE
Backport 08-wasm to v7.3.x

### PR DESCRIPTION
**Pull request by DimitrisJim**
## Description

opening real quick to check CI

closes: #XXXX

### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by damiannolan** [comment]
should this be targeting `08-wasm/release/v0.1.x+ibc-go-v7.3.x-wasmvm-v1.4.x` or is the idea to backport it to `release/v7.3.x` and then cut a new branch from there? slightly confusing

**Comment by crodriguezvega** [comment]
> should this be targeting `08-wasm/release/v0.1.x+ibc-go-v7.3.x-wasmvm-v1.4.x` or is the idea to backport it to `release/v7.3.x` and then cut a new branch from there? slightly confusing

Good catch! Indeed this should target `08-wasm/release/v0.1.x+ibc-go-v7.3.x-wasmvm-v1.4.x`. I will change it now. Hopefully it goes ok!

**Comment by damiannolan** [comment]
> Markdown links seems to fail due to some link in README for the evmos `recovery` module
> 
> https://github.com/cosmos/ibc-go/blob/bef2278049ac46ed97d06c9f94738406355ad739/README.md?plain=1#L76

https://github.com/evmos/evmos/pull/2087

File: modules/light-clients/08-wasm/client/cli/cli.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/client/cli/cli.go:1]: `NewTxCmd` is missing from here with the cli client for store code proposal, and `tx.go` is missing

File: modules/light-clients/08-wasm/go.mod
Comment (damiannolan) [inline modules/light-clients/08-wasm/go.mod:5]: NOTE: another note just to flag removing this local pin before tagging

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:11]: this may be fixed from another backport, but just doing to raise this here as I did on the other backport PR.

```suggestion
	// StoreCode will compile the wasm code, and store the resulting pre-compile
```

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:14]:
```suggestion
	// instantiated many times, and each instance called many times.
```

File: modules/light-clients/08-wasm/keeper/export_test.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/keeper/export_test.go:1]: is this needed anymore?

File: modules/light-clients/08-wasm/keeper/keeper.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/keeper/keeper.go:30]: is this ref needed on the keeper? 

On the main line I think we generally do `ibcwasm.GetVm()`

File: modules/light-clients/08-wasm/keeper/keeper.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/keeper/keeper.go:121]: can we use `ibcwasm.GetVM()`?

File: modules/light-clients/08-wasm/keeper/keeper.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/keeper/keeper.go:132]: ditto

File: modules/light-clients/08-wasm/test_data/ics07_tendermint_cw.wasm.gz
Comment (damiannolan) [inline modules/light-clients/08-wasm/test_data/ics07_tendermint_cw.wasm.gz:1]: think we can remove this now!

File: modules/light-clients/08-wasm/test_data/ics10_grandpa_cw.wasm.gz
Comment (damiannolan) [inline modules/light-clients/08-wasm/test_data/ics10_grandpa_cw.wasm.gz:1]: And this boii

File: modules/light-clients/08-wasm/types/export_test.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/types/export_test.go:13]: Just a note, this isn't needed. There is an exported function which returns this value - `MaxWasmByteSize()`

File: testing/endpoint.go
Comment (damiannolan) [inline testing/endpoint.go:1]: I'm not sure if the changes in this file are relevant or are even used by 08-wasm.

**Review by damiannolan — CHANGES_REQUESTED** [review]
Looks mostly good. I left some notes/comments, I'm sure many can be addressed with backporting the follow ups. 

The biggest thing to me which stood out to me is the missing client CLI for store code

File: modules/light-clients/08-wasm/client/cli/cli.go
Comment (crodriguezvega) [reply modules/light-clients/08-wasm/client/cli/cli.go:1]: We removed it because SDK v0.47 was missing the `SetMsgs` function in `MsgSubmirProposal`. But now that I think about it again, it could be that SDK v0.47.6 now includes it, since I remember Julien was going to back port it to v0.47. I will check and if it's there, then I will either add it back here with a bump of SDK, or I will open a separate PR after this one is merged.

File: modules/light-clients/08-wasm/keeper/keeper.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/keeper.go:30]: Ah, let's leave this for a follow up backport if that resolves it. 

Hard to gauge what exactly is expected in here off the bat

File: modules/light-clients/08-wasm/client/cli/cli.go
Comment (crodriguezvega) [reply modules/light-clients/08-wasm/client/cli/cli.go:1]: I bumped SDK and seems to be working, so I added it back.

Comment (damiannolan) [reply modules/light-clients/08-wasm/client/cli/cli.go:1]: Awesome, nice one @crodriguezvega! Thank you 🙏

**Review by damiannolan — APPROVED** [review]
Approving based on the added back CLI. 

Feel free to apply the godoc suggestions or not, we can follow up by backporting the rest after this goes 🚀

**Comment by damiannolan** [comment]
> > Approving based on the added back CLI.
> 
> We will need though a new release of ibc-go, but I was planning anyway to do it, hopefully next week with the transfer authz improvements and also adding [`08-wasm` here](https://github.com/cosmos/ibc-go/blob/bd169948f83c9c1d28e2fc2a18f089de4f0a6dc1/modules/core/exported/client.go#L26).

Right now, the local pin is okay to move forward with tho? Are you happy to merge this PR now @crodriguezvega? And start following up with more backports, or shall we wait for another review?

**Review by chatton — COMMENTED** [review]
Everything looks good to me, the only thing I'd like to see is a run of the E2Es with images built from this branch. (Even if it's just a manual build and run from `main`).

**Review by chatton — APPROVED** [review]
Good news! 

E2E MsgTransfer Grandpa test passed using the image built from this branch 🥳 

I think that's good enough for me, I don't think we need to test the entire test suite if we have all the unit tests also passing.

```
=== NAME  TestGrandpaTestSuite/TestMsgTransfer_Succeeds_GrandpaContract
    testsuite.go:382: test passed, not uploading logs
    setup.go:210: Pruned 5 volumes, reclaiming approximately 122.2 MB
    setup.go:242: Pruned unused networks: [interchaintest-mzxsyqhm]
--- PASS: TestGrandpaTestSuite (1047.70s)
    --- PASS: TestGrandpaTestSuite/TestMsgTransfer_Succeeds_GrandpaContract (1047.69s)
        --- PASS: TestGrandpaTestSuite/TestMsgTransfer_Succeeds_GrandpaContract/send_successful_IBC_transfer_from_Cosmos_to_Polkadot_parachain (185.62s)
        --- PASS: TestGrandpaTestSuite/TestMsgTransfer_Succeeds_GrandpaContract/send_two_successful_IBC_transfers_from_Polkadot_parachain_to_Cosmos,_first_with_ibc_denom,_second_with_parachain_denom (120.00s)
PASS
ok      github.com/cosmos/ibc-go/e2e/tests/wasm 1047.846s
```

**Pull request merged**
